source/server/hot_restart_impl.cc

```diff
@@ -115,15 +116,15 @@ std::string SharedMemory::version(size_t max_num_stats, size_t max_stat_name_len
 
 HotRestartImpl::HotRestartImpl(Options& options)
     : options_(options), stats_set_options_(sharedMemHashOptions(options.maxStats())),
-      shmem_(SharedMemory::initialize(RawStatDataSet::numBytes(stats_set_options_), options)),
+      shmem_(
+          SharedMemory::initialize(Stats::RawStatDataSet::numBytes(stats_set_options_), options)),
+      allocator_(stats_set_options_, options.restartEpoch() == 0, shmem_.stats_set_data_),
```

Comment:
You will probably need to initialize this below inside the scope where `lock` is created, which is where the stats set used to be initialized.  `allocator_` will probably need to be a unique_ptr as a result.  This is because the lock protects the initialization.  If you move it above to the initializer list, the lock won't be held while it's being constructed.  Side note: are you familiar with the idea of a scope lock (`std::unique_lock<Thread::BasicLockable>`)?

Reply:
This is a good catch. I reworked `allocator_` to be a unique pointer and constructed it not-inline, and changed all references to `allocator_.foo` to `allocator_->foo`.